### PR TITLE
Update dFusionSpec.md

### PR DESCRIPTION
**Pull request by fleupold**
Started rewriting the first paragraphs after feedback from @oliverbeige that the current state is unusable. I am planning to do an entire path over the specification and add sections for account opening/closing.

### CODE REVIEW

File: dFusion/dFusionSpec.md [hunk 1]
```diff
@@ -6,34 +6,69 @@ A specification developed by Gnosis.
 
 
 The following specification uses the snark application (snapp) onchain scaling approach, in order to build a scalable fully decentralized exchange with decentralized order matching. 
-The scalability is enabled storing information only in hashes and allow snarks the manipulation of these hashes in predefined logic gates [cp](https://ethresear.ch/t/on-chain-scaling-to-potentially-500-tx-sec-through-mass-tx-validation/3477).
-In order to allow bigger number of constraints for the snarks, we are planning to use the ideas described in [DIZK](https://www.usenix.org/system/files/conference/usenixsecurity18/sec18-wu.pdf).
+The scalability is enabled storing information only in hashes and allow snarks the manipulation of these hashes in predefined logic gates [[link](https://ethresear.ch/t/on-chain-scaling-to-potentially-500-tx-sec-through-mass-tx-validation/3477)].
```
Comment (bh2smith) [inline dFusion/dFusionSpec.md]:
```suggestion
The scalability is enabled by storing information only in hashes and allows snarks to manipulate these through predefined logic gates [[link](https://ethresear.ch/t/on-chain-scaling-to-potentially-500-tx-sec-through-mass-tx-validation/3477)].
```

File: dFusion/dFusionSpec.md [hunk 2]
```diff
@@ -6,34 +6,69 @@ A specification developed by Gnosis.
...
+
+Anyone can challenge a solution on-chain (while providing a bond and an alternative solution) within a certain time-period after posting.
+Such disputes are resolved by verifying a zkSnark proof on-chain that checks if the provided solution fulfills a number of constraints (specified in detail below).
+If the verification succeeds, the challenger loses their bond and the state is "finalized". 
+If the verification fails the provided alternative state is assumed correct unless challenged within a certain time.
+
+## State stored in the smart contract
+
+For each account, we chain each of ERC20 token balance together and store them as pedersen hash (not merkleized) in the anchor smart contract.
+This "compressed" representation of all account balances is collision resistant and can thus be used  which will store all relevant information for this snapp exchange.
```
Comment (bh2smith) [inline dFusion/dFusionSpec.md]: ... can thus be used <SOMETHING MISSING HERE> which will store all ...

File: dFusion/dFusionSpec.md [hunk 3]
```diff
@@ -6,34 +6,69 @@ A specification developed by Gnosis.
...
+To allow `K` to be small, a bi-map of an accounts public key (on-chain address) to its `accountIndex` will be stored in the anchor contract as well. 
+Accounts will pay "rent" to occupy an active account. The account index can be used to locate a users token balances in the state.
 
-All orders are encoded as limit sell orders: `(accountIndex, fromTokenIndex, toTokenIndex, limitPrice, amount, batchId, signature)`.
-The order should be read in the following way: the user occupying the specified *accountIndex* would like to sell the token *fromTokenIndex* for *toTokenIndex* for at most the *limitPrice* and the *amount* specified.
-The *batchId* and *signature* allow a third party to submit an order on behalf of others (saving gas when batching multiple orders together).
-The user only has to specify which batch their order is valid for and sign all the information with their private key.
+Furthermore, we store a bi-map of token address to its index `0 <= n <= N`, for each token that can be traded on the exchange.
```
Comment (bh2smith) [inline dFusion/dFusionSpec.md]: This is picky, but we may want to consider using `T` for number of tokens and reserve `N` for something else?

File: dFusion/dFusionSpec.md [hunk 4]
```diff
@@ -6,34 +6,72 @@ A specification developed by Gnosis.
...
 2. Transition function from sha to Pedersen hashes (zkSnark)
 3. Finding the batch price: optimization of batch trading surplus (offchain)
 4. Verifying batch price and trade execution (zkSnark)
 5. Processing of pending exits and deposits (zkSnark)
-6. Restart with step 1
+6. Restart with step 0
+
+### Account opening, deposits & withdrawels
+
+// TODO
```
Comment (bh2smith) [inline dFusion/dFusionSpec.md]: I suppose it wouldn't hurt for me to fill in as much of this in as possible before this is merged?

Comment (josojo) [reply dFusion/dFusionSpec.md]: I would keep the deposit and exit descriptions separate from the trade description.
It does not necessarily belong together.

**Pull request merged**